### ISSUE TITLE
Fix performance degredation in multi tower training.

### ISSUE THREAD

**Issue by ppwwyyxx** [ppwwyyxx/tensorpack#5]
When training large models with multi tower, iteration speed can drop significantly after running for a while. Need to investigatet this.

**Comment by ppwwyyxx** [comment]
Found that ILSVRC12 data flow only has a speed of 6it/s while the queue is full. Probably due to serializing and message passing among processes.

**Comment by ppwwyyxx** [comment]
Confirm that pickle is really slow for ILSVRC-size images:
```python
import numpy as np
#import cPickle as pickle
import pickle
import time
d = np.random.rand(32, 299, 299, 3).astype('float32')
print(time.time())
dump = pickle.dumps(d)
print(time.time())
dd = pickle.loads(dump)
print(time.time())
```
cPickle doesn't help. But python3 is much better.

**Comment by ppwwyyxx** [comment]
Confirm the performance issue is due to slow socket communication of multiprocessing.
Will work on a better version.

**Comment by ppwwyyxx** [comment]
If data becomes a bottleneck, PrefetchDataZMQ should be used instead of PrefetchData. Usually several times faster.

**Comment by YixuanLi** [comment]
Hi Yuxin,

I am trying to train a ResNet with two towers, and found the training is very slow (0.18iters/s) on 2 TITAN X. It's taking days to train a CIFAR-10 model. The inference is very fast though (8iters/s). I have cudnn installed and setup appropriately. Do you have any suggestion on how to fix that? 

Thanks!

**Comment by ppwwyyxx** [comment]
That's probably a problem on the environment. Something you can try:
Use a different tensorflow binary. Some previous versions are known to have performance problems.
Don't use cuda 8, there are performance problems reported.
Always use tcmalloc.

**Comment by YixuanLi** [comment]
Thanks for the quick response! That's super helpful. I am using the latest version of tensorflow and cuda7.5. I will look into tcmalloc. Would that make a huge difference (0.18iters/s -> 2iters/s) in terms of performance, or there could be something else going on I should also look into?

**Comment by ppwwyyxx** [comment]
I doubt this alone would make such a huge difference.
What's your speed with one gpu? Ideally it should also be around 2 iter /s.

**Comment by YixuanLi** [comment]
Yeah I wondered that too. Actually even with a single gpu it still remains 0.18iters/s. Any thing else could go wrong? I am following your original setup with customizing anything.

if args.gpu:
        os.environ['CUDA_VISIBLE_DEVICES'] = args.gpu

    config = get_config()
    if args.load:
        config.session_init = SaverRestore(args.load)
    if args.gpu:
        config.nr_tower = len(args.gpu.split(','))
    SyncMultiGPUTrainer(config).train()



Yixuan Li, Ph.D candidate
Cornell University,
Ithaca, NY, USA, 14850
www.cs.cornell.edu/~yli<http://www.cs.cornell.edu/~yli>






On Sep 17, 2016, at 8:32 PM, Yuxin Wu <notifications@github.com<mailto:notifications@github.com>> wrote:


I doubt this alone would make such a huge difference.
What's your speed with one gpu? Ideally it should also be around 2 iter /s.

**Comment by ppwwyyxx** [comment]
That's very strange.
If I ran it with CPU I had 5.54s/it, which is very close to your speed.
Are you sure GPU is used? Can you see your process in `nvidia-smi`? If not then some error must have been printed.

**Comment by YixuanLi** [comment]
Yep the speed has been frustrating me - taking almost 2 days to train cifar-10. I can see my processes on two gpus, each of which uses 4GB memory. I am 97% sure it it running in GPU mode…Does the cuda version matter? I am using cuda4.0 as recommended by tensorflow site. Does the code detect error automatically if cudnn is not being used?

On Sep 17, 2016, at 9:30 PM, Yuxin Wu <notifications@github.com<mailto:notifications@github.com>> wrote:


That's very strange.
If I ran it with CPU I had 5.54s/it, which is very close to your speed.
Are you sure GPU is used? Can you see your process in nvidia-smi? If not then some error must have been printed.

**Comment by YixuanLi** [comment]
I do see these at the beginning:

I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so locally


On Sep 17, 2016, at 9:38 PM, Yixuan Li <yl2363@cornell.edu<mailto:yl2363@cornell.edu>> wrote:

Yep the speed has been frustrating me - taking almost 2 days to train cifar-10. I can see my processes on two gpus, each of which uses 4GB memory. I am 97% sure it it running in GPU mode…Does the cuda version matter? I am using cuda4.0 as recommended by tensorflow site. Does the code detect error automatically if cudnn is not being used?

On Sep 17, 2016, at 9:30 PM, Yuxin Wu <notifications@github.com<mailto:notifications@github.com>> wrote:


That's very strange.
If I ran it with CPU I had 5.54s/it, which is very close to your speed.
Are you sure GPU is used? Can you see your process in nvidia-smi? If not then some error must have been printed.

**Comment by ppwwyyxx** [comment]
Yes, so it shows it's using cudnn. Although cudnn v5 might be better, from their [documents](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/get_started/os_setup.md#requirements).

**Comment by YixuanLi** [comment]
Thanks Yuxin. I’ve configured my environment and made sure tensorflow is all set up appropriately and running in gpu mode. What puzzled me most is the inference still runs very fast whereas the training is drastically slower. Could there be any caveats during training that is causing this? Such as not enough memory allocated, or multi-threading being slow?


100%|#############################################################################################################################################|390/390[35:36<00:00, 0.18it/s]
100%|###############################################################################################################################################|79/79[00:09<00:00, 8.13it/s]



On Sep 17, 2016, at 9:46 PM, Yuxin Wu <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, so it shows it's using cudnn. Although cudnn v5 might be better, from their documents<https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/get_started/os_setup.md#requirements>.

**Comment by YixuanLi** [comment]
Test this for svhn classification: https://github.com/ppwwyyxx/tensorpack/blob/master/examples/svhn-digit-convnet.py, and it yields pretty similar performance as yours (44iters/s).


On Sep 18, 2016, at 9:43 AM, Yixuan Li <yl2363@cornell.edu<mailto:yl2363@cornell.edu>> wrote:

Thanks Yuxin. I’ve configured my environment and made sure tensorflow is all set up appropriately and running in gpu mode. What puzzled me most is the inference still runs very fast whereas the training is drastically slower. Could there be any caveats during training that is causing this? Such as not enough memory allocated, or multi-threading being slow?


100%|#############################################################################################################################################|390/390[35:36<00:00, 0.18it/s]
100%|###############################################################################################################################################|79/79[00:09<00:00, 8.13it/s]



On Sep 17, 2016, at 9:46 PM, Yuxin Wu <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, so it shows it's using cudnn. Although cudnn v5 might be better, from their documents<https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/get_started/os_setup.md#requirements>.

**Comment by ppwwyyxx** [comment]
For the record, it's confirmed to be cudnn v4 problem after working with Yixuan.

**Comment by YixuanLi** [comment]
Confirm. The issue was resolved by upgrading CUDA to 7.5 and cudnn to v5.

On Sep 19, 2016, at 12:07 PM, Yuxin Wu <notifications@github.com<mailto:notifications@github.com>> wrote:


For the record, it's confirmed to be cudnn v4 problem after working with Yixuan.
